tests/AgDatabaseMove.Unit/BackupOrder.cs

```diff
@@ -44,51 +60,158 @@ private static List<BackupMetadata> ListBackups()
+      var duplicate = CloneBackupMetaDataList(listWithStripes);
+      listWithStripes.AddRange(duplicate);
+      return listWithStripes;
+    }
+
+    private static List<BackupMetadata> GetBackupListWithoutFull()
+    {
+      var list = GetBackupList();
+      list.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Full);
+      return list;
+    }
+
+    private static void VerifyListIsAValidBackupChain(IEnumerable<BackupMetadata> backupChain)
+    {
+      var backupChainList = backupChain.ToList();
 
-      var expected = _listBackups.OrderBy(bu => bu.FirstLsn);
+      var listOfFileNames = backupChainList.Select(b => b.PhysicalDeviceName).ToList();
+      Assert.True(listOfFileNames.Count == listOfFileNames.Distinct().Count());
+
+      var orderedBackups = CloneBackupMetaDataList(backupChainList)
+        .OrderBy(b => b.LastLsn)
+        // 'PhysicalDeviceName' ordering doesn't matter but need it here to simplify this check for LSN ordering
+        .ThenBy(b => b.PhysicalDeviceName).ToList();
+      Assert.True(backupChainList.SequenceEqual(orderedBackups, new BackupMetadataEqualityComparer()));
+
+      // first one has to be full
+      var full = backupChainList.FirstOrDefault();
+      Assert.NotNull(full);
+      Assert.True(full.BackupType == BackupFileTools.BackupType.Full);
+
+      // removing all the striped files of the full backup
+      backupChainList.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Full &&
+                                     b.CheckpointLsn == full.CheckpointLsn);
+
+      var otherFullBackups = backupChainList.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Full);
+      Assert.Equal(0, otherFullBackups);
+
+      // next one can be diff or log
+      var currentBackup = backupChainList.FirstOrDefault();
+      if(currentBackup == null) {
+        return;
+      }
+
+      Assert.True(currentBackup.DatabaseBackupLsn == full.CheckpointLsn);
+
+      if(currentBackup.BackupType == BackupFileTools.BackupType.Diff) {
+
+        // removing all the striped files of the diff backup
+        backupChainList.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Diff &&
+                                       b.FirstLsn == currentBackup.FirstLsn && b.LastLsn == currentBackup.LastLsn);
+
+        var otherDiffBackups = backupChainList.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Diff);
+        Assert.Equal(0, otherDiffBackups);
+      }
+
+      var prevBackup = currentBackup;
+      backupChainList.ForEach(b => {
+        Assert.True(b.BackupType == BackupFileTools.BackupType.Log &&
+                    (
+                      (prevBackup.FirstLsn == b.FirstLsn && prevBackup.LastLsn == b.LastLsn) ||
+                      prevBackup.LastLsn == b.FirstLsn)
+                    );
+        prevBackup = b;
+      });
 
-      Assert.Equal<IEnumerable>(backupChain.OrderedBackups, expected);
     }
 
-    [Fact]
-    public void MissingLink()

```

Comment:
@alexirion10 not sure how to test this - any ideas?
I don't think the current testing strategy for this is a complete one - i.e just asserting that the length of the output is not the length of the expected output 

I think we should ideally have a way to detect this in our logic and throw an exception if it occurs - then test for the exception here.

One potential idea I have is to see if there were other log backups still left in the list we didn't include in the chain - and go through their LSNs to see if they should have been in the chain - hence identifying the missing link

But is a bit complicated to do - so I'm wondering if it's best for another PR?
What are your thoughts?

Reply:
no no you don't need to do all of that.  The existing test is a good one

1. Get a full backup chain.
2. Remove 1 of the items from the middle of the chain
Assert 1 has all backups in the chain up to the one removed.  

We rely on the SQL tech for chaining LSN's. Once the chain is broken, I'm not aware of how to find a missing link.

Reply:
ahh I see, sure will place that test back in

>We rely on the SQL tech for chaining LSN's

I guess this is also to do with scope right - this module assumes LSNs are done correctly?